File: lastStoneWeight.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class Solution
{
public:
  int lastStoneWeight(vector<int> &stones)
  {
    int num1, num2; // num1 >= num2 (most heaviest stones in the heap)
    priority_queue<int> pq;
    for (int i = 0; i < stones.size(); ++i)
    {
      pq.push(stones[i]);
    }
    if (pq.empty())
    {
      return 0;
    }
    while (pq.size() > 1)
    {
      num1 = pq.top();
      pq.pop();
      num2 = pq.top();
      pq.pop();
      if (num1 > num2)
      {
        pq.push(num1 - num2);
      }
    }
    return pq.empty() ? 0 : pq.top();
  }
};

File: LastStoneWeight.py
from queue import PriorityQueue

class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        pq = PriorityQueue()
        for stone in stones:
            pq.put(-stone)  # negate to act as a max heap since it is a minHeap

        while pq.qsize() >= 2:
            stone1 = -pq.get()
            stone2 = -pq.get()
            if stone1 > stone2:
                pq.put(-(stone1 - stone2))

        return -pq.get() if not pq.empty() else 0


"""
another approach using sort
"""

class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        stones.sort()
        while(len(stones)>1):
            stone1 = stones.pop()
            stone2 = stones.pop()
            if stone1 > stone2:
                stones.append(stone1-stone2)
                stones.sort()
        return stones[0] if stones else 0





